lighthouse-core/gather/computed/first-interactive.js

```diff
@@ -0,0 +1,206 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const ComputedArtifact = require('./computed-artifact');
+const TracingProcessor = require('../../lib/traces/tracing-processor');
+
+const LONG_TASK_THRESHOLD = 50;
+
+const MAX_TASK_CLUSTER_DURATION = 250;
+const MIN_TASK_CLUSTER_PADDING = 1000;
+const MIN_TASK_CLUSTER_FMP_DISTANCE = 5000;
+
+const MAX_QUIET_WINDOW_SIZE = 5000;
+const TRACE_BUSY_MSG = 'trace was busy the entire time';
+
+/**
+ * @fileoverview This artifact identifies the time the page is "first interactive" as defined below
+ * @see https://docs.google.com/document/d/1GGiI9-7KeY3TPqS3YT271upUVimo-XiL5mwWorDUD4c/edit#
+ *
+ * First Interactive marks the first moment when a website is minimally interactive:
+ *    > Enough (but maybe not all) UI components shown on the screen are interactive
+ *      DISCLAIMER: This is assumed by virtue of the fact that the CPU is idle, actual event
+ *      listeners are not examined, server-side rendering and extreme network latency can dupe this
+ *      definition.
+ *    > The page responds to user input in a reasonable time on average, but it’s ok if this
+ *      response is not always immediate.
+ *
+ * First Interactive is defined as the first period after FMP of N-seconds that has no bad task
+ * clusters.
+ *
+ *    > t = time in seconds since FMP
+ *    > N = f(t) = 4 * e^(-0.045 * t) + 1
+ *      5 = f(0) = 4 + 1
+ *      3 ~= f(15) ~= 2 + 1
+ *      1 ~= f(∞) ~= 0 + 1
+ *    > a "bad task cluster" is a cluster of 1 or more long tasks with no more than 1s of idle time
+ *      between each task that does one of the following
+ *        > Starts within the first 5s after FMP.
+ *        > Spans more than 250ms from the start of the earliest task in the cluster to the end of the
+ *          latest task in the cluster.
+ *
+ * If this timestamp is earlier than DOMContentLoaded, use DOMContentLoaded as firstInteractive.
+ */
+class FirstInteractive extends ComputedArtifact {
+  get name() {
+    return 'FirstInteractive';
+  }
+
+  /**
+   * @param {number} t The time passed since FMP in miliseconds.
+   * @return {number}
+   */
+  static getRequiredWindowSizeInMs(t) {
+    const tInSeconds = t / 1000;
+    const exponentiationComponent = Math.exp(-.045 * tInSeconds);
```

Comment:
we should probably use the canonical `-Math.log(2) / 15` instead of `-.045` since we're going to want to compare against other implementations. The difference is small but will definitely be noticeable when e.g. comparing analysis of the same saved trace